chore(Dropdown): updated deprecated usage to new dropdown

thatblindgeye opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Closes #8839 

Pages from the surge that were updated (some with notes, other notes after this list):

- [Action list single group](https://patternfly-react-pr-8868.surge.sh/components/action-list#action-list-single-group)
- [Breadcrumb with dropdown](https://patternfly-react-pr-8868.surge.sh/components/breadcrumb#with-dropdown) - uses a recreation of the badge toggle
- [Card examples](https://patternfly-react-pr-8868.surge.sh/components/card) that have a kebab and [horizontal card grid demo](https://patternfly-react-pr-8868.surge.sh/components/card/react-demos#horizontal-card-grid) 
- [Datalist examples](https://patternfly-react-pr-8868.surge.sh/components/data-list) that have a kebab, and the demo
- [Datetime picker](https://patternfly-react-pr-8868.surge.sh/components/date-and-time/date-and-time-picker) - the time toggle (with clock icon) no longer has the bottom border since I'm using the "plain" MenuToggle variant to remove the caret icon. Could be a Core update or we could introduce a new MenuToggle variant ("controlled"? basically something that keeps the bottom border but removes the caret icon)
- [Dual list selector complex options and actions](https://patternfly-react-pr-8868.surge.sh/components/dual-list-selector#using-more-complex-options-with-actions)
- [Hint examples](https://patternfly-react-pr-8868.surge.sh/components/hint)
- [Input group with dropdown](https://patternfly-react-pr-8868.surge.sh/components/input-group#with-dropdown)
- [Masthead demos](https://patternfly-react-pr-8868.surge.sh/components/masthead/react-demos)
- [Menu with drilldown breadcrumb](https://patternfly-react-pr-8868.surge.sh/components/menus/menu#with-drilldown-breadcrumbs)
- [Modal with dropdown](https://patternfly-react-pr-8868.surge.sh/components/modal#with-dropdown)
- [Navigation demos](https://patternfly-react-pr-8868.surge.sh/components/navigation/react-demos) - Horizontal, Horizontal nav with horizontal subnav, Manual nav
- [Notification drawer](https://patternfly-react-pr-8868.surge.sh/components/notification-drawer) - first 2 examples and both demos; these demos + cypress demos need to be looked into as keyboard handling doesn't work as expected (might be due to NotificationDrawerListItem's internal onKeyDown firing)
- [Page demos](https://patternfly-react-pr-8868.surge.sh/components/page/react-demos)
- [Primary detail pattern](https://patternfly-react-pr-8868.surge.sh/patterns/primary-detail) - Primary-detail card view, Primary-detail data list in card
- [Tabs and tables demo](https://patternfly-react-pr-8868.surge.sh/components/tabs/react-demos#tables-and-tabs)
- [Toolbar with filters](https://patternfly-react-pr-8868.surge.sh/components/toolbar#with-filters), [Toolbar stacked](https://patternfly-react-pr-8868.surge.sh/components/toolbar#stacked-example), and [Toolbar demos](https://patternfly-react-pr-8868.surge.sh/components/toolbar/react-demos)
- [Treeview with action items](https://patternfly-react-pr-8868.surge.sh/components/tree-view#with-action-items) - cannot currently navigate between kebab menu items via keyboard. This *may* be due to both TreeView and the new Dropdown's internal Menu components both using KeyboardHandler
- [Alert group demos](https://patternfly-react-pr-8868.surge.sh/components/alert-group/react-demos)
- [Card view pattern](https://patternfly-react-pr-8868.surge.sh/patterns/card-view) - see note below

Other notes:

- Demos that use selectable cards + action kebabs (Card view, primary details, etc): clicking the kebab causes the cards' selection to occur
- Looks like the wrapper `div` that is rendered around Dropdown's Popper prevents MenuToggle from being full height when `isFullHeight` is passed (we may not need that wrapper anymore since moving forward we're defaulting to an "inline" appendTo, which that wrapper div was just to give the Popper something to append to)
- The new Dropdown inside of PageHeader components (non-Masthead) doesn't have the correct styling

<details>
<summary>Original comment regarding badge toggle</summary>

Wanted to put a draft up to get some opinion on the Menu Toggle with badge implementation (seen in the Breadcrumb with dropdown example here).

With the deprecated Dropdown, there was a separate BadgeToggle sub-component that was used, but MenuToggle doesn't have that. What I did for now in this PR is just pass Badge component as a child to MenuToggle (rather than using the `badge` prop), and set it to a `plainText` variant. I figured right now that would be the simplest way to show how a consumer could use this new composable implementation.

Alternatively we could maybe implement a new prop in MenuToggle that would create a "BadgeToggle", but one possible issue is whether that would get confusing since we already have an `icon` and `badge` prop on MenuToggle. Some ideas:

- Introduce a new "badge" variant, along with "plain", "default", etc. that would set the markup and styling similar to how BadgeToggle did
- Add a new prop (isBadgeToggle?) that would probably do the same as above
- Show how to use the menu classes to create a custom badge toggle, basically leave it up to the consumer (probably not great since it'd involve hardcoding classes in?)

There may need some Core followup as I did notice the breadcrumb items with a dropdown weren't centered right. Changing the `.pf-c-breadcrumb__item` "align-items" property to `center` instead of `baseline` may help, but I haven't looked into whether that would cause any other issues.

</details>

<!-- Are there any upstream issues or separate issues you need to reference? -->
**Additional issues**:

patternfly-build commented:
Preview: https://patternfly-react-pr-8868.surge.sh

mcarrano commented:
@mceledonia @mmenestr @andrew-ronaldson would like to get your opinion on this one. See the question posed by @thatblindgeye in the description above. The gist of this is that by deprecating the old dropdown, we loose the badge variant that allows us to make a badge into a toggle. We could add that to Menu Toggle, but there's a question how useful this is. This was a feaure that was requested for condensing a long breadcrumb control. Not aware that it's used anyplace else. Here are two links where you can compare the existing vs. new implementation:

Existing: https://www.patternfly.org/v4/components/breadcrumb#with-dropdown

Updated: https://patternfly-react-pr-8868.surge.sh/components/breadcrumb#with-dropdown

Note the changed positioning of the caret. I'm OK with the updated implementation but also wanted to get your opinion. Don't recall where this is being used, but if you are aware, I can also check with that team as they may need to do some rework to update to PatternFly 5.

thatblindgeye commented:
@mcarrano Nicole pointed me towards a demo that has a badge toggle using MenuToggle similar to the old Dropdown BadgeToggle: [composable dropdown variants](http://patternfly-react-v5.surge.sh/components/menus/custom-menus/#composable-dropdown-variants)

So I can actually implement that in this update, unless another approach is agreed on. From what I have seen so far, two areas where we're using a badge toggle like this in examples is the Breadcrumb example and the Menu Drilldown with breadcrumb example.

kmcfaul commented:
> * Looks like the wrapper `div` that is rendered around Dropdown's Popper [...]
You should be able to remove the wrapping div if you pass `triggerRef` and `popperRef` to the Popper.


For the TreeView action menu, I thought the keyboard handler should prevent propagation up to the outer handler (so if the focus is in the Dropdown that handler should run and not the TreeView) but maybe that needs some tweaking.

mcarrano reviewed (changes requested):
@thatblindgeye Glad you were able to work out the issue with the badge style dropdown. But I am seeing a spacing difference between this and the current implementation:

Current:
![Screenshot 2023-04-12 at 4 30 50 PM](https://user-images.githubusercontent.com/6619860/231577686-7bd34a63-2bcf-4e1f-a797-c5fa4dff3518.png)

This PR:
![Screenshot 2023-04-12 at 4 30 40 PM](https://user-images.githubusercontent.com/6619860/231577779-f31afc8a-30e2-4f9e-a100-1a1d7c418dc0.png)

Can that be fixed?

mcarrano commented:
@thatblindgeye Took some time this morning to review other places affected by this change. Here are additional problems I found...

In all full page demos, the user menu is not rendering correctly. See below.

![Screenshot 2023-04-13 at 9 17 19 AM](https://user-images.githubusercontent.com/6619860/231773905-95293040-e17a-455f-955c-d4ba388c475e.png)

Should look like this:

![Screenshot 2023-04-13 at 9 30 27 AM](https://user-images.githubusercontent.com/6619860/231774361-c13b7886-485a-4443-9cc3-f7c4d9b5c7aa.png)

I'm not entirely comfortable with effect of this on the Date & Time picker as it now feels like the time widget is no longer a part of the control.

![Screenshot 2023-04-13 at 9 14 52 AM](https://user-images.githubusercontent.com/6619860/231774854-661a6270-262f-4b0c-bd4a-a35fa788a94a.png)

I wonder if we'd be better off adding back the border even if it means it means we need to keep the caret. 

![Screenshot 2023-04-13 at 9 14 40 AM](https://user-images.githubusercontent.com/6619860/231775453-ff5c3124-b025-4d73-89f6-15f7a6cd1528.png)

@mmenestr @andrew-ronaldson what do you think?

thatblindgeye commented:
@mcarrano I should be able to resolve the user menu issue based on Katie's comment above.

Will wait for Andrew and/or Margot's input on the timepicker (can also make updates to the badge toggle depending what you think of my latest comment above)

mcarrano commented:
@thatblindgeye regarding fixes to the badge dropdown, I'd be OK with either approach. Although option 2 is perhaps a more robust way to do this? Maybe see what others think.

thatblindgeye commented:
@mcarrano Made some updates to help resolve the spacing issue on the badge toggle as well as the dropdown used in full page demos for the username (should now be full height as it was before).

@mmenestr For the Timepicker, I removed the variant so that it matches the second screenshot Matt shared for now. Should we open a followup issue to resolve how we ultimately want that timepicker toggle to look?

mcarrano commented:
I took at look at your changes @thatblindgeye . The breadcrumb update looks good. For the Date & Time picker, can we fix this so that the clock icon does not get cut off?

@mmenestr agree about the Date & Time demo. Would you mind opening an issue to reconsider that? It's definitely beyond the scope of this PR.

mmenestr commented:
@thatblindgeye Thank you for making that update !
@mcarrano I opened up the following issue to address this: https://github.com/patternfly/patternfly-design/issues/1256

mcarrano commented:
> If the spacing isn't too bad I'll push that change, since it may make more sense to pass only an icon to the icon prop.

@thatblindgeye it does feel like a bit of extra space, but I guess this is OK as this is only really a demo.

mcarrano commented:
Thanks for cleaning up the date & time picker @thatblindgeye . The look of the user menus in the full page demos is still not correct. Should be using the Avatar in front of the username like this...

![Screenshot 2023-04-14 at 1 31 06 PM](https://user-images.githubusercontent.com/6619860/232116686-0bd6c4cd-a7ea-458f-855f-2ae61359facf.png)

mcarrano reviewed (approved):
All looks good now. Thanks @thatblindgeye !

tlabaj commented:
It looks like there is a spacing issue with some of the card example.  There is a large space after the separator and the next dropdown item.
![image](https://user-images.githubusercontent.com/39532947/232589553-af3deb24-44de-4328-985f-f671baa2be36.png)

tlabaj commented:
> * [Datetime picker](https://patternfly-react-pr-8868.surge.sh/components/date-and-time/date-and-time-picker) - the time toggle (with clock icon) no longer has the bottom border since I'm using the "plain" MenuToggle variant to remove the caret icon. Could be a Core update or we could introduce a new MenuToggle variant ("controlled"? basically something that keeps the bottom border but removes the caret icon)

What do you thinnk @mcarrano @srambach

mcarrano commented:
@tlabaj regarding the Date & Time example, we worked through some alternatives and landed on a compromise as expressed in this comment: https://github.com/patternfly/patternfly-react/pull/8868#issuecomment-1508425193

The extra spacing in the card dropdown is a good catch. However I'm not seeing this when I view the same example. Thoughts @thatblindgeye ?

thatblindgeye commented:
@mcarrano I pushed an update to resolve the issue @tlabaj mentioned so you might have seen the examples with those updates. There were just some rogue commas being rendered causing that extra space to be taken up

nicolethoen commented:
I'm noticing that the Dropdown's MenuToggle is not receiving focus back after selecting a Dropdown item. That is what is happening in the deprecated dropdown.

And upon further inspection. It looks like the same is true for the Select. We will likely need to address this. https://github.com/patternfly/patternfly-react/issues/8985

thatblindgeye commented:
Created https://github.com/patternfly/patternfly-react/issues/9012 as a followup to look into the bug that occurs in some full page demos.

tlabaj reviewed (commented):
LGTM

srambach commented:
I ran through all your links at the top - found some things that seem off:

There's a small overlap in the focus ring in the date and time picker
<img width="477" alt="image" src="https://user-images.githubusercontent.com/19825616/235458861-d44a9d9d-1df1-4c2a-ad03-1e6bcdd0cdbd.png">

Looks like there's a problem rendering the Composable Structure example https://patternfly-react-pr-8868.surge.sh/components/dual-list-selector/#composable-structure

Drilldown breadcrumbs have 2 different formats - on the first one the caret is inside and the third it's outside
https://patternfly-react-pr-8868.surge.sh/components/menus/menu/#with-drilldown-breadcrumbs

The navigation demos username still seem out of alignment
<img width="493" alt="image" src="https://user-images.githubusercontent.com/19825616/235460711-79590ec5-f0d6-401b-9072-576caf4fc137.png">

Notification drawer is having a weird flash of scrollbar and the menu won't stay when using the keyboard

On the Primary-detail demos I'm getting "ReferenceError: SelectVariant is not defined"

In Tabs - tables and tabs, the menu isn't dropping down in the normal position; it's higher than usual https://patternfly-react-pr-8868.surge.sh/components/tabs/react-demos/tables-and-tabs/

On some of the demos I catch the menu flash in the wrong place before showing up correctly.  I don't remember ever seeing that before?
https://user-images.githubusercontent.com/19825616/235464415-c1d19f6d-48ea-4389-8732-39d265009df8.mp4

thatblindgeye commented:
@srambach 

- Date and time picker: it looks like that issue could be resolved by React updating the markup to include the `pf-c-input-group__item` element (this could possibly covered by #8953 ?)
- Composable Structure: that should just be a code block to show the intended structure of the composable component
- Menu with drilldown breadcrumb: updated
- Navigation demos: that might be tied to my original comment: "The new Dropdown inside of PageHeader components (non-Masthead) doesn't have the correct styling". @tlabaj would we want to do a followup to update the affected Nav demos to not use PageHeader since it's deprecated now?
- Notification Drawer: noted the issue, I can create a followup if we want to get this in, otherwise I can try digging a little more into it
- PrimaryDetail: updated the import to the deprecated usage since it'll need to be updated using the new Select and new functionality introduced in #8992 (can make a followup issue for this)
- Tabs demo: that looks to be a pre-existing issue, can also open a followup for this

The last item I opened up a followup to investigate that further.

nicolethoen commented on packages/react-integration/demo-app-ts/src/components/demos/NotificationDrawerDemo/NotificationDrawerGroupsDemo.tsx:
```diff
@@ -88,32 +86,37 @@ export class GroupsNotificationDrawerDemo extends React.Component<
       <NotificationDrawer>
         <NotificationDrawerHeader count={4} onClose={this.onDrawerClose}>
-          <DropdownDeprecated
+          <Dropdown
```
Same again `onOpenChange`

thatblindgeye commented:
@nicolethoen updated those tests since it could be useful down the line

srambach reviewed (approved):
This one looks good with the follow up issues.
